File: pruebas.py
import random
import string
import time
import sys
sys.setrecursionlimit(100000)

# Tiempos de QS y BS
tmps_qs_asc, tmps_bs_asc = [], []
tmps_qs_desc, tmps_bs_desc = [], []
tmps_qs_rand, tmps_bs_rand = [], []

def tam_arr_pruebas():
    tams = []
    for i in range(20):
        tam = int(input("Ingrese el tamanio " + str(i + 1) + ": "))
        if tam == 0:
            break
        tams.append(tam)
    return tams

def pruebas():
    tams = tam_arr_pruebas()
    n = len(tams)

    print("\nTamaños de arrays:")
    imprimir_arrreglo(tams)

    # Ascendente
    print("\nEjecutando pruebas en arreglos ascendentes...")
    for i in range(n):
        tam = tams[i]
        arr = gen_arr(tam, "ascendente")

        # QuickSort
        inicio = time.time()
        qs_ord = quick_sort(arr, asc=True)
        fin = time.time()
        tiempo_qs = fin - inicio

        # BubbleSort
        inicio = time.time()
        bs_ord = bubble_sort(arr)
        fin = time.time()
        tiempo_bs = fin - inicio

        #print(f"\nARREGLO NRO {i + 1}")
        #print("==================")
        #imprimir_resultados(arr, qs_ord, bs_ord)

        tmps_qs_asc.append(tiempo_qs)
        tmps_bs_asc.append(tiempo_bs)

    # Descendente
    print("\nEjecutando pruebas en arreglos descendentes...")
    for i in range(n):
        tam = tams[i]
        arr = gen_arr(tam, "descendente")

        # QuickSort
        inicio = time.time()
        qs_ord = quick_sort(arr, asc=False)
        fin = time.time()
        tiempo_qs = fin - inicio

        # BubbleSort
        inicio = time.time()
        bs_ord = bubble_sort(arr)
        fin = time.time()
        tiempo_bs = fin - inicio

        #print(f"\nARREGLO NRO {i + 1}")
        #print("==================")
        #imprimir_resultados(arr, qs_ord, bs_ord)

        tmps_qs_desc.append(tiempo_qs)
        tmps_bs_desc.append(tiempo_bs)

    # Desordenado
    print("\nEjecutando pruebas en arreglos desordenados...")
    for i in range(n):
        tam = tams[i]
        arr = gen_arr(tam, "desordenado")

        # QuickSort
        inicio = time.time()
        qs_ord = quick_sort(arr, asc=True)
        fin = time.time()
        tiempo_qs = fin - inicio

        # BubbleSort
        inicio = time.time()
        bs_ord = bubble_sort(arr)
        fin = time.time()
        tiempo_bs = fin - inicio

        #print(f"\nARREGLO NRO {i + 1}")
        #print("==================")
        #imprimir_resultados(arr, qs_ord, bs_ord)

        tmps_qs_rand.append(tiempo_qs)
        tmps_bs_rand.append(tiempo_bs)
    
    return tams, n

def gen_arr(tam, escenario):
    arr = generar_arreglo(tam)
    if escenario == "ascendente":
        arr_asc = quick_sort(arr, asc=True)
        return arr_asc
    elif escenario == "descendente":
        arr_desc = quick_sort(arr, asc=False)
        return arr_desc
    elif escenario == "desordenado":
        return arr

# Generar una lista de caracteres aleatorios
def generar_arreglo(tam):
    return [random.choice(string.ascii_lowercase) for _ in range(tam)]

def imprimir_resultados(arr_gen, arr_bs, arr_qs):
    print("\nArreglo generado:")
    imprimir_arrreglo(arr_gen)

    print("\nArreglo ordenado por Bubble Sort:")
    imprimir_arrreglo(arr_bs)

    print("\nArreglo ordenado por Quick Sort:")
    imprimir_arrreglo(arr_qs)


def imprimir_arrreglo(arr):
    for i in range(len(arr)):
        print(arr[i], end=' ')
    print()

def imprimir_tiempos(tipo_orden, tiempos_bs, tiempos_qs, tamanos, nro_arrays):
    print(f"\n{tipo_orden}:")
    for i in range(nro_arrays):
        print(f"Tamaño {tamanos[i]}: Bubble Sort: {tiempos_bs[i]:.6f}, Quick Sort: {tiempos_qs[i]:.6f}")

def guardar_tiempos(nombre_archivo):
    try:
        with open(nombre_archivo, "w") as archivo:
            for i in range(len(tamanos)):
                archivo.write(f"{tamanos[i]} {tmps_bs_asc[i]:.6f} {tmps_qs_asc[i]:.6f} "
                            f"{tmps_bs_desc[i]:.6f} {tmps_qs_desc[i]:.6f} "
                            f"{tmps_bs_rand[i]:.6f} {tmps_qs_rand[i]:.6f}\n")
        print(f"\nLos tiempos se han guardado exitosamente en '{nombre_archivo}'")
    except IOError:
        print(f"Error al abrir el archivo '{nombre_archivo}' para guardar los tiempos.")

def quick_sort(arr, asc=True):
    arr_ordenado = arr.copy()
    quick_sort_recursivo(arr_ordenado, 0, len(arr_ordenado) - 1, asc)
    return arr_ordenado

def quick_sort_recursivo(arr, inicio, fin, asc):
    if inicio < fin:
        pivote = particionar(arr, inicio, fin, asc)
        quick_sort_recursivo(arr, inicio, pivote - 1, asc)
        quick_sort_recursivo(arr, pivote + 1, fin, asc)

def particionar(arr, inicio, fin, asc):
    x = arr[fin]
    i = inicio - 1
    for j in range(inicio, fin):
        if asc:
            if arr[j] <= x:
                i += 1
                arr[i], arr[j] = arr[j], arr[i] #? intercambio
        
        else:
            if arr[j] >= x:
                i += 1
                arr[i], arr[j] = arr[j], arr[i] #? intercambio
    
    arr[i + 1], arr[fin] = arr[fin], arr[i + 1] #? intercambio
    return i + 1

def bubble_sort(arr):
    arr_ordenado = arr.copy()
    n = len(arr_ordenado)
    for i in range(n):
        for j in range(0, n - i - 1):
            if arr_ordenado[j] > arr_ordenado[j + 1] :
                arr_ordenado[j], arr_ordenado[j + 1] = arr_ordenado[j + 1], arr_ordenado[j]
    return arr_ordenado

tamanos, nro_arrays = pruebas()

print("\nTiempos de ejecución (en segundos):")

imprimir_tiempos("Ascendente", tmps_bs_asc, tmps_qs_asc, tamanos, nro_arrays)
imprimir_tiempos("Descendente", tmps_bs_desc, tmps_qs_desc, tamanos, nro_arrays)
imprimir_tiempos("Aleatorio", tmps_bs_rand, tmps_qs_rand, tamanos, nro_arrays)

# Guardar tiempos en un archivo de texto
guardar_tiempos("tiempos_python.txt")

File: pruebas.c
#include <stdio.h>
#include <stdlib.h>   // Para malloc y free
#include <string.h>  // Para memcpy
#include <time.h>    // Para medir los tiempos

// Prototipos de funciones en NASM
extern char gen_char_aleatorio(void);
extern void gen_caracteres(char* arr, int tam);
extern void quick_sort(char* arr, int inicio, int fin, int bandera);
extern void bubble_sort(char* arr, int tam);
extern void guardarTams(void);
extern long nroArrays;
extern long tamanos[20];

double tmps_qs_asc[20], tmps_bs_asc[20];
double tmps_qs_desc[20], tmps_bs_desc[20];
double tmps_qs_rand[20], tmps_bs_rand[20];

void imprimirArr(char* arr, size_t tam) {
    for (size_t i = 0; i < tam; i++) {
        printf("%c ", arr[i]);
    }
    printf("\n");
}

void imprimirTam(long* arr, long tam) {
    printf("\nTamaños de arrays:\n");
    for (long i = 0; i < tam; i++) {
        printf("%ld ", arr[i]);
    }
    printf("\n");
}

void imprimirResultados(char* ptrArr, char* ptrArrBS, char* ptrArrQS, size_t tamArr) {
    printf("\nArreglo generado:\n");
    imprimirArr(ptrArr, tamArr);

    printf("\nArreglo ordenado con Bubble Sort:\n");
    imprimirArr(ptrArrBS, tamArr);

    printf("\nArreglo ordenado con Quick Sort:\n");
    imprimirArr(ptrArrQS, tamArr);
}

void imprimirTiempos(char* tipoOrden, double* tiemposBS, double* tiemposQS, long* tamanos, int nroArrays) {
    printf("\n%s:\n", tipoOrden);
    for (int i = 0; i < nroArrays; i++) {
        printf("Tamaño %ld: Bubble Sort: %.6f, Quick Sort: %.6f\n", tamanos[i], tiemposBS[i], tiemposQS[i]);
    }
}

double medirTiempo(void (*algOrdenamiento)(char*, int), char* arr, int tam) {
    clock_t inicio, fin;
    inicio = clock();
    algOrdenamiento(arr, tam);
    fin = clock();
    return ((double) (fin - inicio)) / CLOCKS_PER_SEC;
}

void quickSortWrap(char* arr, int tam) {
    quick_sort(arr, 0, tam - 1, 1);
}

void guardarTiempos(const char* nombreArchivo) {
    FILE* archivo = fopen(nombreArchivo, "w");

    if (!archivo) {
        printf("Error al abrir el archivo para guardar los tiempos.\n");
        return;
    }

    for (int i = 0; i < nroArrays; i++) {
        // Guardar los tiempos en el archivo (tamaño, BubbleSort, QuickSort)
        fprintf(archivo, "%ld %.6f %.6f %.6f %.6f %.6f %.6f\n", tamanos[i],
                tmps_bs_asc[i], tmps_qs_asc[i], tmps_bs_desc[i], tmps_qs_desc[i], tmps_bs_rand[i], tmps_qs_rand[i]);
    }

    fclose(archivo);
    printf("\nLos tiempos se han guardado exitosamente en %s\n", nombreArchivo);
}

void pruebas() {
    char *ptrArr, *ptrArrBS, *ptrArrQS;
    size_t tamArr;
    int i;

    // Prueba para orden ascendente
    printf("\nEjecutando pruebas en arreglos ascendentes...\n");

    for (i = 0; i < nroArrays; i++) {
        // Asignar tamaño del arreglo actual
        tamArr = tamanos[i];

        // Asignar memoria para los arreglos
        ptrArr = malloc(tamArr);
        ptrArrBS = malloc(tamArr);
        ptrArrQS = malloc(tamArr);

        if (!ptrArr || !ptrArrBS || !ptrArrQS) {
            printf("Error al asignar memoria\n");
            return;
        }

        // Generar caracteres aleatorios
        gen_caracteres(ptrArr, tamArr);

        // Generar array en orden ascendente
        quick_sort(ptrArr, 0, tamArr - 1, 1);

        // Copiar el arreglo generado a los arreglos de burbuja y quicksort
        memcpy(ptrArrBS, ptrArr, tamArr);
        memcpy(ptrArrQS, ptrArr, tamArr);

        tmps_bs_asc[i] = medirTiempo(bubble_sort, ptrArrBS, tamArr);
        tmps_qs_asc[i] = medirTiempo(quickSortWrap, ptrArrQS, tamArr);

        // Imprimir resultados
        printf("\nARREGLO NRO %d \n", (i + 1));
        printf("===============\n");
        imprimirResultados(ptrArr, ptrArrBS, ptrArrQS, tamArr);

        // Liberar la memoria asignada
        free(ptrArr);
        free(ptrArrBS);
        free(ptrArrQS);
    }

    // Prueba para orden descendente
    printf("\nEjecutando pruebas en arreglos descendentes...\n");

    for (i = 0; i < nroArrays; i++) {
        // Asignar tamaño del arreglo actual
        tamArr = tamanos[i];

        // Asignar memoria para los arreglos
        ptrArr = malloc(tamArr);
        ptrArrBS = malloc(tamArr);
        ptrArrQS = malloc(tamArr);

        if (!ptrArr || !ptrArrBS || !ptrArrQS) {
            printf("Error al asignar memoria\n");
            return;
        }

        // Generar caracteres aleatorios
        gen_caracteres(ptrArr, tamArr);

        // Generar array en orden descendente
        quick_sort(ptrArr, 0, tamArr - 1, -1);

        // Copiar el arreglo generado a los arreglos de burbuja y quicksort
        memcpy(ptrArrBS, ptrArr, tamArr);
        memcpy(ptrArrQS, ptrArr, tamArr);

        tmps_bs_desc[i] = medirTiempo(bubble_sort, ptrArrBS, tamArr);
        tmps_qs_desc[i] = medirTiempo(quickSortWrap, ptrArrQS, tamArr);

        // Imprimir resultados
        printf("\nARREGLO NRO %d \n", (i + 1));
        printf("===============\n");
        imprimirResultados(ptrArr, ptrArrBS, ptrArrQS, tamArr);

        // Liberar la memoria asignada
        free(ptrArr);
        free(ptrArrBS);
        free(ptrArrQS);
    }

    // Prueba para orden desordenado
    printf("\nEjecutando pruebas en arreglos desordenados...\n");

    for (i = 0; i < nroArrays; i++) {
        // Asignar tamaño del arreglo actual
        tamArr = tamanos[i];

        // Asignar memoria para los arreglos
        void *ptrArr = malloc(tamArr);
        void *ptrArrBS = malloc(tamArr);
        void *ptrArrQS = malloc(tamArr);

        if (!ptrArr || !ptrArrBS || !ptrArrQS) {
            printf("Error al asignar memoria\n");
            return;
        }

        // Generar array con caracteres aleatorios
        gen_caracteres(ptrArr, tamArr);

        // Copiar el arreglo generado a los arreglos de burbuja y quicksort
        memcpy(ptrArrBS, ptrArr, tamArr);
        memcpy(ptrArrQS, ptrArr, tamArr);

        tmps_bs_rand[i] = medirTiempo(bubble_sort, ptrArrBS, tamArr);
        tmps_qs_rand[i] = medirTiempo(quickSortWrap, ptrArrQS, tamArr);

        // Imprimir resultados
        printf("\nARREGLO NRO %d \n", (i + 1));
        printf("===============\n");
        imprimirResultados(ptrArr, ptrArrBS, ptrArrQS, tamArr);

        // Liberar la memoria asignada
        free(ptrArr);
        free(ptrArrBS);
        free(ptrArrQS);
    }
}

int main() {
    srand(time(NULL));  // Inicializar la semilla para rand()

    guardarTams();

    imprimirTam(tamanos, nroArrays);
    pruebas();

    // Mostrar tiempos de ejcucion
    printf("\nTiempos de ejecución (en segundos):\n");

    imprimirTiempos("Ascendente", tmps_bs_asc, tmps_qs_asc, tamanos, nroArrays);
    imprimirTiempos("Descendente", tmps_bs_desc, tmps_qs_desc, tamanos, nroArrays);
    imprimirTiempos("Aleatorio", tmps_bs_rand, tmps_qs_rand, tamanos, nroArrays);

    // Guardar tiempos de ejecución en un archivo
    guardarTiempos("tiempos_asm.txt");

    return 0;
}